README.rst
Search String Parser
====================

.. list-table::
    :stub-columns: 1

    * - docs
      - |docs|
    * - tests
      - | |travis|
    * - package
      - |version| |wheel| |supported-versions| |supported-implementations|

.. |docs| image:: https://readthedocs.org/projects/searchstringparser/badge/?version=latest
    :target: http://searchstringparser.readthedocs.org/en/latest/?badge=latest
    :alt: Documentation Status

.. |travis| image:: https://travis-ci.org/AGHerwig/searchstringparser.svg?branch=master
    :alt: Travis-CI Build Status
    :target: https://travis-ci.org/AGHerwig/searchstringparser

.. |version| image:: https://img.shields.io/pypi/v/searchstringparser.svg?style=flat
    :alt: PyPI Package latest release
    :target: https://pypi.org/project/searchstringparser

.. |wheel| image:: https://img.shields.io/pypi/wheel/searchstringparser.svg?style=flat
    :alt: PyPI Wheel
    :target: https://pypi.org/project/searchstringparser

.. |supported-versions| image:: https://img.shields.io/pypi/pyversions/searchstringparser.svg?style=flat
    :alt: Supported versions
    :target: https://pypi.org/project/searchstringparser

.. |supported-implementations| image:: https://img.shields.io/pypi/implementation/searchstringparser.svg?style=flat
    :alt: Supported implementations
    :target: https://pypi.org/project/searchstringparser

Parse a more general search syntax to conform with a particular SQL dialect.

Currently, this is implemented using ply_ with a general lexer and a parser for
generating PostgreSQL-specific search queries.

* Free software: BSD license

.. _ply: http://www.dabeaz.com/ply/ply.html

Installation
============

::

    pip install searchstringparser

Documentation
=============

https://searchstringparser.readthedocs.org/

Development
===========

To run the all tests run::

    tox


CHANGELOG.rst

Changelog
=========

0.3.0 (2018-10-04)
-----------------------------------------

* Fix a bug in the token parser.
* Update the infrastructure.

0.2.3 (2015-09-29)
-----------------------------------------

* Publish on PyPi

0.2.2 (2015-09-29)
-----------------------------------------

* Complete documentation

0.2.1 (2015-09-28)
-----------------------------------------

* Add more helpful error messages

0.2.0 (2015-09-28)
-----------------------------------------

* Add complete integration with Travis, Appveyor, Coveralls, and Read the Docs
* Increase test coverage to 100%
* Make source compatible with Python 2.6 - 3.4

0.1.1 (2015-09-21)
-----------------------------------------

* Fix GeneralSearchStringLexer regex
* Fix PostgreSQLTextSearchParser rules
* Add first set of tests

0.1.0 (2015-09-16)
-----------------------------------------

* Initial class layout



test_general.py

import sys
sys.path.insert(0, "src")

from searchstringparser import (GeneralSearchStringLexer,
        PostgreSQLTextSearchParser)

def test_general():
    lexer = GeneralSearchStringLexer()
    query = "stuff 'in quotes \\' is   good \"stuff'"
    lexer.test(query)

def test_postgre():
    parser = PostgreSQLTextSearchParser()
    query = "stuff 'in quotes \\' is   good \"stuff'"
    print(parser.parse(query))

if __name__ == "__main__":
    test_general()
    test_postgre()



tests/lexer/test_general.py
# -*- coding: utf-8 -*-


"""
=======================================
General Search String Token Lexer Tests
=======================================

:Authors:
    Moritz Emanuel Beber
:Date:
    2015-09-18
:Copyright:
    Copyright |c| 2015, Max Plank Institute for Molecular Genetics,
    all rights reserved.
:File:
    test_general.py

.. |c| unicode: U+A9
"""


from __future__ import absolute_import

import io

import pytest

from searchstringparser import GeneralSearchStringLexer


@pytest.fixture
def lexer():
    return GeneralSearchStringLexer()


@pytest.fixture
def general_output():
    return io.open("tests/lexer/general.out").readlines()


class TestGeneralSearchStringLexer(object):
    @pytest.mark.parametrize("query,expected", [
        ("stuff", ("WORD",)),
        ("stuff stuff", ("WORD", "WORD")),
        ("stuff  \t \r \nstuff", ("WORD", "WORD")),
        ("stuff*", ("WORD", "WILDCARD")),
        ("-stuff", ("NOT", "WORD")),
        ("- stuff", ("NOT", "WORD")),
        ("~stuff", ("NOT", "WORD")),
        ("~ stuff", ("NOT", "WORD")),
        ("!stuff", ("NOT", "WORD")),
        ("! stuff", ("NOT", "WORD")),
        ("not stuff", ("NOT", "WORD")),
        ("NOT stuff", ("NOT", "WORD")),
        ("stuff & stuff", ("WORD", "AND", "WORD")),
        ("stuff && stuff", ("WORD", "AND", "WORD")),
        ("stuff and stuff", ("WORD", "AND", "WORD")),
        ("stuff AND stuff", ("WORD", "AND", "WORD")),
        ("stuff | stuff", ("WORD", "OR", "WORD")),
        ("stuff || stuff", ("WORD", "OR", "WORD")),
        ("stuff or stuff", ("WORD", "OR", "WORD")),
        ("stuff OR stuff", ("WORD", "OR", "WORD")),
        ("(stuff OR stuff)", ("LPAREN", "WORD", "OR", "WORD", "RPAREN")),
        (
            "(good stuff) (more stuff)",
            ("LPAREN", "WORD", "WORD", "RPAREN", "LPAREN", "WORD", "WORD",
             "RPAREN")
        ),
        ("'stuff'", ("QUOTE", "SYMBOL", "QUOTE")),
        ("\"stuff\"", ("QUOTE", "SYMBOL", "QUOTE")),
        (
            "\"stuff goes on\"",
            ("QUOTE", "SYMBOL", "SPACE", "SYMBOL", "SPACE", "SYMBOL", "QUOTE")
        ),
        ("\"stuff '\"", ("QUOTE", "SYMBOL", "SPACE", "LITERAL_QUOTE", "QUOTE")),
        (
            "\"stuff 'goes\\\" on\"",
            ("QUOTE", "SYMBOL", "SPACE", "LITERAL_QUOTE", "SYMBOL",
             "LITERAL_QUOTE", "SPACE", "SYMBOL", "QUOTE")
        ),
        ("order foo", ("WORD", "WORD")),
        ("foo or order", ("WORD", "OR", "WORD")),
        ("foo OR order", ("WORD", "OR", "WORD")),
        ("foo | order", ("WORD", "OR", "WORD")),
        ("foo || order", ("WORD", "OR", "WORD")),
        ("foo & andrew", ("WORD", "AND", "WORD")),
        ("foo && andrew", ("WORD", "AND", "WORD")),
        ("foo and andrew", ("WORD", "AND", "WORD")),
        ("foo AND andrew", ("WORD", "AND", "WORD")),
    ])
    def test_token(self, lexer, query, expected):
        lexer.input(query)
        tokens = tuple(tok.type for tok in lexer)
        assert tokens == expected

    @pytest.mark.parametrize("query,illegal", [
        ("stuff", None),
        ("%stuff;", (["%", ";"], [0, 6]))
    ])
    def test_illegal(self, lexer, query, illegal):
        lexer.input(query)
        # consume input s.t. tokens are generated
        for _ in lexer:
            pass
        assert lexer.get_illegal() == illegal

    @pytest.mark.parametrize("query,lines", [
        ("stuff", (0,)),
        ("stuff stuff", (0, 6)),
        ("stuff  \t \r \nstuff", (0, 12)),
        ("stuff*", (0, 15)),
        ("-stuff", (16, 1)),
        ("- stuff", (16, 2)),
        ("~stuff", (17, 1)),
        ("~ stuff", (17, 2)),
        ("!stuff", (18, 1)),
        ("! stuff", (18, 2)),
        ("not stuff", (19, 4)),
        ("NOT stuff", (20, 4)),
        ("stuff & stuff", (0, 21, 8)),
        ("stuff && stuff", (0, 22, 9)),
        ("stuff and stuff", (0, 23, 10)),
        ("stuff AND stuff", (0, 24, 10)),
        ("stuff | stuff", (0, 25, 8)),
        ("stuff || stuff", (0, 26, 9)),
        ("stuff or stuff", (0, 27, 9)),
        ("stuff OR stuff", (0, 28, 9)),
        ("(stuff OR stuff)", (30, 1, 29, 10, 32)),
        ("(good stuff) (more stuff)", (30, 35, 6, 33, 31, 36, 37, 34)),
        ("'stuff'", (38, 42, 39)),
        ("\"stuff\"", (40, 42, 41)),
        ("\"stuff goes on\"", (40, 42, 46, 43, 47, 44, 45)),
        ("\"stuff '\"", (40, 42, 46, 48, 49)),
        ("stuff", (0,)),
        ("%stuff;", (1, 50, 51))
    ])
    def test_print(self, lexer, capsys, general_output, query, lines):
        lexer.print_tokens(query)
        (out, err) = capsys.readouterr()
        expected = "".join(general_output[i] for i in lines)
        print(query)
        assert out == expected
        assert err == ""


tests/parser/test_postgresql.py
# -*- coding: utf-8 -*-


"""
====================================
PostgreSQL Search Token Parser Tests
====================================

:Authors:
    Moritz Emanuel Beber
:Date:
    2015-09-18
:Copyright:
    Copyright |c| 2015, Max Plank Institute for Molecular Genetics,
    all rights reserved.
:File:
    test_postgresql.py

.. |c| unicode: U+A9
"""


from __future__ import absolute_import

import pytest

from searchstringparser import PostgreSQLTextSearchParser


@pytest.fixture
def parser():
    return PostgreSQLTextSearchParser()


class TestGeneralSearchStringLexer(object):
    @pytest.mark.parametrize("query,expected", [
        ("stuff", "stuff"),
        ("stuff stuff", "stuff & stuff"),
        ("stuff  \t \r \nstuff", "stuff & stuff"),
        ("stuff stuff", "stuff & stuff"),
        ("stuff*", "stuff:*"),
        ("-stuff", "!stuff"),
        ("- stuff", "!stuff"),
        ("~stuff", "!stuff"),
        ("~ stuff", "!stuff"),
        ("!stuff", "!stuff"),
        ("! stuff", "!stuff"),
        ("not stuff", "!stuff"),
        ("NOT stuff", "!stuff"),
        ("stuff & stuff", "stuff & stuff"),
        ("stuff && stuff", "stuff & stuff"),
        ("stuff and stuff", "stuff & stuff"),
        ("stuff AND stuff", "stuff & stuff"),
        ("stuff | stuff", "stuff | stuff"),
        ("stuff || stuff", "stuff | stuff"),
        ("stuff or stuff", "stuff | stuff"),
        ("stuff OR stuff", "stuff | stuff"),
        ("(stuff OR stuff)", "(stuff | stuff)"),
        ("(good stuff) (more stuff)", "(good & stuff) & (more & stuff)"),
        ("'stuff'", "'stuff'"),
        ("\"stuff\"", "'stuff'"),
        ("\"stuff goes on\"", "'stuff goes on'"),
        ("\"stuff '\"", "'stuff '''"),
        ("\"stuff 'goes\\\" on\"", "'stuff ''goes'' on'"),
        ("(good stuff) || (more 'almost \"decent\"' stuff)",
            "(good & stuff) | (more & 'almost ''decent''' & stuff)"),
    ])
    def test_parse(self, parser, query, expected):
        output = parser.parse(query)
        assert output == expected

    @pytest.mark.parametrize("query,expected,illegal,illegal_pos", [
        ("stuff", "stuff", None, None),
        ("%stuff;", "stuff", ("%", ";"), (0, 6))
    ])
    def test_illegal(self, parser, query, expected, illegal, illegal_pos):
        output = parser.parse(query)
        assert output == expected
        illegal_ = parser.get_illegal()
        if illegal_ is not None:
            assert illegal_[0] == illegal
            assert illegal_[1] == illegal_pos

    @pytest.mark.parametrize("query,message", [
        ("stuff & & stuff", "LexToken(AND,'&',1,8)"),
        ("'stuff", "Unclosed quote at position 0."),
        (
            "(stuff",
            "1 mismatched parentheses! Last opening parenthesis at position 0."
        ),
        (
            "stuff)",
            "1 mismatched parentheses! Last closing parenthesis at position 5."
        ),
    ])
    def test_errors(self, parser, query, message):
        with pytest.raises(SyntaxError) as excinfo:
            parser.parse(query)
        assert str(excinfo.value) == message


src/searchstringparser/parser/postgresql.py
# -*- coding: utf-8 -*-


"""
==============================
PostgreSQL Search Token Parser
==============================

:Authors:
    Moritz Emanuel Beber
:Date:
    2015-09-16
:Copyright:
    Copyright |c| 2015, Max Plank Institute for Molecular Genetics,
    all rights reserved.
:File:
    postgresql.py

.. |c| unicode: U+A9
"""


from __future__ import absolute_import

import ply.yacc as yacc

from ..lexer import GeneralSearchStringLexer


__all__ = ["PostgreSQLTextSearchParser"]


class PostgreSQLTextSearchParser(object):
    """
    This parser implements the following rules using the tokens generated by an
    appropriate lexer. The goal is to generate a string for PostgreSQL full text
    search that conforms with the syntax understood by the function `tsquery or
    to_tsquery
    <http://www.postgresql.org/docs/9.4/static/datatype-textsearch.html#DATATYPE-TSQUERY>`_.

    The following rules are implemented which generate the correct query string.

    ::

        expression : expression expression
                   | expression AND expression
                   | expression OR expression
                   | NOT expression
                   | LPAREN expression RPAREN
                   | QUOTE term QUOTE
                   | WORD WILDCARD
                   | WORD

        term : term SPACE term
               | term term
               | LITERAL_QUOTE
               | SYMBOL
               
    """  # noqa

    precedence = (
        ("left", "OR"),
        ("left", "AND", "SPACE"),
        ("right", "NOT"),
        ("left", "WILDCARD")
    )

    def __init__(self, lexer=None, **kw_args):
        """Parser instantiation.

        Parameters
        ----------
        lexer : ply.lex (optional)
            Any ply.lex lexer instance and generates the tokens listed in the
            rules. The default uses a GeneralSearchStringLexer instance.
        kw_args :
            Keyword arguments are passed to the ply.yacc.yacc call.

        """
        super(PostgreSQLTextSearchParser, self).__init__()
        self.lexer = GeneralSearchStringLexer() if lexer is None else lexer
        self.tokens = self.lexer.tokens
        self.parser = yacc.yacc(module=self, **kw_args)

    def parse(self, query, **kw_args):
        """Parse any string input according to the rules.

        Parameters
        ----------
        query : str
            A string expected to conform with the search query syntax.
        kw_args :
            Keyword arguments are passed on to the ply.yacc.yacc.parse call.

        Returns
        -------
        str
            A string that can directly be passed to the PostgreSQL functions
            `tsquery or to_tsquery <http://www.postgresql.org/docs/9.4/static/datatype-textsearch.html#DATATYPE-TSQUERY>`_.
            
        """  # noqa
        self.lexer._invalid = list()
        self.lexer._invalid_pos = list()
        return self.parser.parse(query, lexer=self.lexer, **kw_args)

    def get_illegal(self):
        """Inspect encountered illegal characters.

        Returns
        -------
        None
            If no illegal characters occurred.
        Tuple
            A pair of lists that contain the illegal characters and
            the positions where they occurred.
        """
        self.lexer.get_illegal()

    def p_error(self, p):
        if p is None:
            if self.lexer.parens_level > 0:
                raise SyntaxError(
                    "{0:d} mismatched parentheses! Last opening"
                    " parenthesis at position {1:d}.".format(
                            self.lexer.parens_level,
                            self.lexer.last_lparens
                    ))
            elif self.lexer.last_quote is not None:
                raise SyntaxError("Unclosed quote at position {0:d}.".format(
                    self.lexer.last_quote
                ))
#            else:  # there should be no other EOF syntax errors
#                raise SyntaxError("Syntax error at EOF!")
        if self.lexer.parens_level < 0:
            raise SyntaxError(
                "{0:d} mismatched parentheses! Last closing"
                " parenthesis at position {1:d}.".format(
                        abs(self.lexer.parens_level),
                        self.lexer.last_rparens
                ))
        else:
            raise SyntaxError(str(p))

    def p_expression_space(self, p):
        """expression : expression expression %prec AND"""
        p[0] = "{0} & {1}".format(p[1], p[2])

    def p_expression_and(self, p):
        """expression : expression AND expression"""
        p[0] = "{0} & {1}".format(p[1], p[3])

    def p_expression_or(self, p):
        """expression : expression OR expression"""
        p[0] = "{0} | {1}".format(p[1], p[3])

    def p_expression_unot(self, p):
        """expression : NOT expression"""
        p[0] = "!{0}".format(p[2])

    def p_expression_parens(self, p):
        """expression : LPAREN expression RPAREN"""
        p[0] = "({0})".format(p[2])

    def p_expression_quoted(self, p):
        """expression : QUOTE term QUOTE"""
        p[0] = "'{0}'".format(p[2])

    def p_expression_word_wildcard(self, p):
        """expression : WORD WILDCARD"""
        p[0] = "{0}:*".format(p[1])

    def p_expression_word(self, p):
        """expression : WORD"""
        p[0] = p[1]

    def p_term_term_space_term(self, p):
        """term : term SPACE term"""
        p[0] = "{0} {1}".format(p[1], p[3])

    def p_term_term_term(self, p):
        """term : term term %prec WILDCARD"""
        p[0] = "{0}{1}".format(p[1], p[2])

    def p_term_literal_quote(self, p):
        """term : LITERAL_QUOTE"""
        p[0] = "''"

    def p_term_symbol(self, p):
        """term : SYMBOL"""
        p[0] = p[1]


src/searchstringparser/lexer/general.py
# -*- coding: utf-8 -*-


"""
=================================
General Search String Token Lexer
=================================

:Authors:
    Moritz Emanuel Beber
:Date:
    2015-09-16
:Copyright:
    Copyright |c| 2015, Max Plank Institute for Molecular Genetics,
    all rights reserved.
:File:
    general.py

.. |c| unicode: U+A9
"""


from __future__ import absolute_import, print_function

import ply.lex as lex


__all__ = ["GeneralSearchStringLexer"]


class GeneralSearchStringLexer(object):
    """
    """

    states = (
        ("quoting", "exclusive"),
    )
    # List of token names.   This is always required
    tokens = (
       "WORD",
       "WILDCARD",
       "NOT",
       "AND",
       "OR",
       "LPAREN",
       "RPAREN",
       "SYMBOL",
       "LITERAL_QUOTE",
       "SPACE",
       "QUOTE"
    )

    # A string containing ignored characters (spaces and tabs)
    t_ignore = " \t\r\n\f\v"  # whitespace (interpreted literally not as regex)
    # Regular expression rules for simple tokens
    t_WORD = r"\w+"
    t_WILDCARD = r"\*"
    t_NOT = r"-|~|!|not\b|NOT\b"
    t_AND = r"&{1,2}|and\b|AND\b"
    t_OR = r"\|{1,2}|or\b|OR\b"
    # rules for 'quoting' state
    t_quoting_ignore = ""
    t_quoting_SYMBOL = r"([^'\"\s\\]|\\(?!'|\"))+"  # anything but other tokens

    def __init__(self, illegal="ignore", **kw_args):
        """A composite class of a ply.lex.lex.

        This is the setup step necessary before you can iterate over the tokens.

        Parameters
        ----------
        illegal : {'record', 'ignore', 'error'} (optional)
            Action to be taken when illegal characters are encountered. The
            default is to record them but continue without regarding them.
        kw_args :
            Keyword arguments are passed to the ply.lex.lex call.
        """
        super(GeneralSearchStringLexer, self).__init__()
# TODO: pick between different error handling methods
        self.lexer = lex.lex(module=self, **kw_args)
        self.parens_level = 0
        self.last_lparens = 0
        self.last_rparens = 0
        self.last_quote = None
        self._quote_start = None
        self._invalid = None
        self._invalid_pos = None

    def __getattr__(self, attr):
        return getattr(self.lexer, attr)

    def __iter__(self):
        return (tok for tok in iter(self.lexer.token, None))

    def input(self, data):
        """Add a new string to the lexer.

        This is the setup step necessary before you can iterate over the tokens.

        Parameters
        ----------
        data : str
            Any string.
        """
        self.lexer.push_state("INITIAL")
        self.parens_level = 0
        self.last_lparens = 0
        self.last_rparens = 0
        self.last_quote = None
        self._quote_start = None
        self._invalid = list()
        self._invalid_pos = list()
        self.lexer.input(data)

    def get_illegal(self):
        """Return encountered illegal characters.

        Returns
        -------
        None
            If no illegal characters occurred.
        Tuple
            A pair of lists that contain the illegal characters and
            the positions where they occurred.
        """
        if not self._invalid:
            return None
        return self._invalid, self._invalid_pos

    # inspect output
    def print_tokens(self, data):
        """Print all tokens in a string.

        First iterates through all tokens found and prints them to sys.stdout.
        Then prints illegal characters if any occurred.

        Parameters
        ----------
        data : str
            Any string.
        """
        self.input(data)
        for tok in self:
            print(tok)
        if self._invalid:
            print("Invalid character(s):", ", ".join(self._invalid))
            print("at position(s):",
                  ", ".join([str(x) for x in self._invalid_pos]))

    # Error handling rule
    def t_ANY_error(self, t):
        self._invalid.append(t.value[0])
        self._invalid_pos.append(t.lexpos)
        t.lexer.skip(1)

    def t_LPAREN(self, t):
        r"\("
        self.parens_level += 1
        self.last_lparens = t.lexer.lexpos - 1
        return t

    def t_RPAREN(self, t):
        r"\)"
        self.parens_level -= 1
        self.last_rparens = t.lexer.lexpos - 1
        return t

    # quoting state
    def t_quoting_LITERAL_QUOTE(self, t):
        r"\\'|\\\""
        return t

    def t_QUOTE(self, t):
        r"'|\""
        self._quote_start = t.lexer.lexpos
        self.last_quote = t.lexer.lexpos - 1
        t.lexer.push_state("quoting")
        return t

    def t_quoting_QUOTE(self, t):
        r"'|\""
        # handle different quote inside quote
        if t.value != t.lexer.lexdata[self._quote_start - 1]:
            t.type = "LITERAL_QUOTE"
            return t
        self.last_quote = None
        t.lexer.pop_state()
        return t

    def t_quoting_SPACE(self, t):
        r"\s+"
        return t
